### PR TITLE
Event Cart: convert to use standard payment forms

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Convert the event cart to use standard payment forms and supported interfaces. Builds on #17339.

Before
----------------------------------------
Very old, unsupported code. Only a single payment processor supported and some don't work (eg. Stripe). Lot's of custom code to create contribution/participants/lineitems and payment made before contribution created.

After
----------------------------------------
Checkout:
image
(The pay now button is a sneak-preview of the upcoming google/apple/payment request support in stripe).

Settings:
image

Technical Details
----------------------------------------
This is a major refactor of the event cart checkout following it's move to an extension. We switch to using standard billingblock functionality and support an independent set of payment processors via settings. (Previously this was guessed from the events in the cart and only one processor could be used).
This switches event cart to use Order API + payment propertybag making this probably some of the most up to date code in CiviCRM!

Comments
----------------------------------------
@eileenmcnaughton @artfulrobot This is still a work in progress because I've not gone through and verified that receipting is working properly (and may switch to Payment.sendconfirmation for this). But you'll be pleased to see it uses Order API to create the contribution, participants and lineitems. And then uses payment propertybag before calling doPayment.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I'm working on trying to get the review queue under control so I'm closing out these PRs that acheive the same function closed as open.

**Comment by eileenmcnaughton** [comment]
(obviously it might make sense to rebase the branch intermittently to see what might still be worth bringing over from here - )

**Comment by eileenmcnaughton** [comment]
I should note there are 2 significant blockers on this work based on a quick skim
1) the DAO move - there are technical problems to solve there for new installs before we can look at moving the DAO
2) the Use of the property bag - I think we have figured out that we are quite far apart on our vision for what we want to achieve for the property bag & that needs resolving before we do more PRs on it (although I don't have much appetite for getting back into that at the moment)

**Comment by mattwire** [comment]
Can be tested by via https://github.com/mjwconsult/civicrm-eventcart

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
```
CRM_Financial_Form_PaymentFormsTest::testEventPaymentForms
Deprecated function PropertyBag array access for core property 'amount', use getAmount().

/home/jenkins/bknix-dfl/build/core-17886-7noia/web/sites/all/modules/civicrm/CRM/Core/Error.php:1053
/home/jenkins/bknix-dfl/build/core-17886-7noia/web/sites/all/modules/civicrm/CRM/Core/Error.php:1042
/home/jenkins/bknix-dfl/build/core-17886-7noia/web/sites/all/modules/civicrm/Civi/Payment/PropertyBag.php:154
/home/jenkins/bknix-dfl/build/core-17886-7noia/web/sites/all/modules/civicrm/CRM/Core/Payment.php:1362
/home/jenkins/bknix-dfl/build/core-17886-7noia/web/sites/all/modules/civicrm/ext/eventcart/CRM/Event/Cart/Form/Checkout/Payment.php:340
/home/jenkins/bknix-dfl/build/core-17886-7noia/web/sites/all/modules/civicrm/tests/phpunit/CRM/Financial/Form/PaymentFormsTest.php:93
/home/jenkins/bknix-dfl/build/core-17886-7noia/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:226
/home/jenkins/bknix-dfl/extern/phpunit7/phpunit7.phar:615
```

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
```
CRM_Financial_Form_PaymentFormsTest::testEventPaymentForms
Deprecated function Legacy property name 'is_recur', use Canonical property name 'isRecur'.

/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/CRM/Core/Error.php:1053
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/CRM/Core/Error.php:1042
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/Civi/Payment/PropertyBag.php:269
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/Civi/Payment/PropertyBag.php:138
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/CRM/Core/Payment/AuthorizeNet.php:127
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/CRM/Core/Payment.php:1388
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/ext/eventcart/CRM/Event/Cart/Form/Checkout/Payment.php:340
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/tests/phpunit/CRM/Financial/Form/PaymentFormsTest.php:93
/home/jenkins/bknix-dfl/build/core-17886-1eqpf/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:226
/home/jenkins/bknix-dfl/extern/phpunit7/phpunit7.phar:615
```

File: CRM/Core/Payment/AuthorizeNet.php [hunk 1]
```diff
@@ -113,6 +113,10 @@ protected function supportsRecurContributionsForPledges() {
    * @throws \Civi\Payment\Exception\PaymentProcessorException
    */
   public function doPayment(&$params, $component = 'contribute') {
+    if ($params instanceof \Civi\Payment\PropertyBag) {
```
Comment (eileenmcnaughton) [inline CRM/Core/Payment/AuthorizeNet.php:116]: I know this is your work in progress PR & we are not supposed to be looking at it but I stumbled here from your other PR & saw this - we can't actually have something that requires the payment processors to add something like this in what we actually merge

Comment (mattwire) [reply CRM/Core/Payment/AuthorizeNet.php:116]: I've put it there for now but we can also put it in the eventcart code just before calling `doPayment()` which would be a nice way of allowing us to pass a propertyBag to a paymentprocessor without requiring any modifications on the paymentprocessor side. Additionally a paymentprocessor could explicitly choose to switch ON legacy warnings once it's suitably converted.

Comment (eileenmcnaughton) [reply CRM/Core/Payment/AuthorizeNet.php:116]: Yep - not requiring any modifications on the processor side is the pre-requisite for us to start passing out propertyBag from any core functions

**Comment by demeritcowboy** [comment]
This has merge conflicts and has been sitting for several months with a needs-work tag so going to close but feel free to reopen.

**Pull request closed without merging**